src/Symfony/Component/HttpKernel/Controller/ArgumentResolver.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\Controller;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\ControllerMetadata\Argument\ArgumentMetadataFactoryInterface;
+
+/**
+ * Responsible for the resolving of arguments passed to an action.
+ *
+ * @author Iltar van der Berg <kjarli@gmail.com>
+ */
+final class ArgumentResolver implements ArgumentResolverInterface
+{
+    /**
+     * @var ArgumentMetadataFactoryInterface
+     */
+    private $argumentMetadataFactory;
+
+    /**
+     * @var ArgumentValueResolverInterface[]
+     */
+    private $argumentValueResolvers;
+
+    /**
+     * @param ArgumentMetadataFactoryInterface $argumentMetadataFactory
+     * @param ArgumentValueResolverInterface[] $argumentValueResolvers
+     */
+    public function __construct(ArgumentMetadataFactoryInterface $argumentMetadataFactory = null, array $argumentValueResolvers = array())
+    {
+        $this->argumentMetadataFactory = $argumentMetadataFactory;
+        $this->argumentValueResolvers = $argumentValueResolvers;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getArguments(Request $request, $controller)
+    {
+        $arguments = array();
+
+        foreach ($this->argumentMetadataFactory->createArgumentMetadata($controller) as $metadata) {
+            $isResolved = false;
+            foreach ($this->argumentValueResolvers as $resolver) {
+                if ($resolver->supports($request, $metadata)) {
+                    $resolved = $resolver->getValue($request, $metadata);
+
+                    // variadic is a special case, always being the last and being an array
+                    if (is_array($resolved) && $metadata->isVariadic()) {
+                        return array_merge($arguments, $resolved);
+                    }
```

Comment:
That would introduce the following:
 - DX goes down slightly as for every return an array would have to be returned while this would only be for 1 specific case in the core. It feels weird if you have to return an array for 1 value.
 - array_merge per iteration

This could be solved by the following:
```php
$resolved = $resolver->getValue($request, $metadata);
$arguments[] = is_array($resolved) ? $resolved : array($resolved);

// and finish off by merging it
return call_user_func_array('array_merge', $arguments);
```

This would mean a slight performance nudge for 1 specific case though.